### PR TITLE
Allow blockquote bubbles to fill the message bubble

### PR DESCRIPTION
**Pull request by pixlwave**
Use a custom layout to prevent them from unnecessarily widening the bubble.

I hope the code is well enough documented to describe how this is done, please add comments where unclear and I'll add more documentation.

IMG_DBBE4B12F03B-1

### CODE REVIEW

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift [hunk 1]
```diff
@@ -111,10 +209,31 @@ struct FormattedBodyText_Previews: PreviewProvider {
...
             FormattedBodyText(text: "Some plain text that's not an attributed component.")
+                .previewBubble()
+            FormattedBodyText(text: "Some plain text that's not an attributed component. This one is really long.")
+                .previewBubble()
         }
+        .padding()
+    }
+}
+
+private struct PreviewBubbleModifier: ViewModifier {
```
Comment (stefanceriu) [inline ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift:224]: Nice! 👍

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift [hunk 2]
```diff
@@ -111,10 +209,31 @@ struct FormattedBodyText_Previews: PreviewProvider {
...
+    func body(content: Content) -> some View {
+        content
+            .padding(timelineStyle == .bubbles ? 8 : 0)
+            .background(timelineStyle == .bubbles ? Color.element.systemGray6 : nil)
+            .cornerRadius(timelineStyle == .bubbles ? 12 : 0)
+    }
+}
+
+private extension View {
+    func previewBubble() -> some View {
```
Comment (stefanceriu) [inline ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift:236]: Not pariticularly important in this case but I wonder if we should name modifiers according to what they actually do e.g. `wrapInPreviewBubble` 🤔

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift [hunk 3]
```diff
@@ -17,43 +17,134 @@
...
+    let spacing: CGFloat
+    
+    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
+        guard !subviews.isEmpty else { return .zero }
+        
+        // Calculate the natural size using the regular text and non-greedy blockquote bubbles.
+        let layoutSubviews = subviews.filter { $0.priority != .blockquoteRenderPriority }
+        
+        let subviewSizes = layoutSubviews.map { $0.sizeThatFits(proposal) }
+        let subviewHeight = subviewSizes.map(\.height).reduce(0, +)
```
Comment (stefanceriu) [inline ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift:47]: Is this maybe `totalHeight`

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift [hunk 4]
```diff
@@ -17,43 +17,134 @@
...
+        let width = subviewSizes.map(\.width).reduce(0, max)
```
Comment (stefanceriu) [inline ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift:48]: and this `maxWidth`?

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift [hunk 5]
```diff
@@ -17,43 +17,134 @@
 import Foundation
 import SwiftUI
 
+/// Layout priority constants for `FormattedBodyText`. These priorities are abused within
+/// `FormattedBodyTextBubbleLayout` to create the layout we would like. They aren't
+/// used in the expected way that SwiftUI would normally use layout priorities.
+private extension Double {
```
Comment (stefanceriu) [inline ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift:23]: Why does this need to be an extension on Double as opposed to just any other enum?

**Review by stefanceriu — APPROVED** [review]
Some comments inline but also talked about this offline, looks good to me 👍

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift (hunk 2 above)
Comment (pixlwave) [reply ElementX/Sources/Screens/RoomScreen/View/Timeline/FormattedBodyText.swift:236]: Discussed this one and decided its less in fitting with Apple's `background(…)`, `cornerRadius(…)` etc.

**Pull request merged**